db/db_versioned.go

```diff
@@ -80,50 +95,195 @@ func (b *BoltDBVersioned) Stop(ctx context.Context) error {
+		}).serialize(), "failed to create metadata")
+	} else {
+		if len(key) != int(vn.keyLen) {
+			return errors.Wrapf(ErrInvalid, "invalid key length, expecting %d, got %d", vn.keyLen, len(key))
+		}
+	}
+	// check key's metadata
+	km, err := b.checkKey(ns, key)
+	if err != nil {
+		return err
+	}
+	km, exit := km.updateWrite(version, value)
+	if exit {
+		return nil
+	}
+	buf.Put(ns, append(key, 0), km.serialize(), fmt.Sprintf("failed to put key %x's metadata", key))
+	buf.Put(ns, versionedKey(key, version), value, fmt.Sprintf("failed to put key %x", key))
+	return b.db.WriteBatch(buf)
 }
 
 // Get retrieves the most recent version
-func (b *BoltDBVersioned) Get(ns string, version uint64, key []byte) ([]byte, error) {
+func (b *BoltDBVersioned) Get(version uint64, ns string, key []byte) ([]byte, error) {
 	if !b.db.IsReady() {
 		return nil, ErrDBNotStarted
 	}
-	// TODO: implement Get
-	return nil, nil
+	// check key's metadata
+	km, err := b.checkNamespaceAndKey(ns, key)
+	if err != nil {
+		return nil, err
+	}
+	hitLast, err := km.updateRead(version)
+	if err != nil {
+		return nil, errors.Wrapf(err, "key = %x", key)
+	}
+	if hitLast {
+		return km.lastWrite, nil
+	}
+	return b.get(version, ns, key)
+}
+
+func (b *BoltDBVersioned) get(version uint64, ns string, key []byte) ([]byte, error) {
+	// construct the actual key = key + version (represented in 8-bytes)
+	// and read from DB
+	key = versionedKey(key, version)
+	var value []byte
+	err := b.db.db.View(func(tx *bolt.Tx) error {
+		bucket := tx.Bucket([]byte(ns))
+		if bucket == nil {
+			return errors.Wrapf(ErrBucketNotExist, "bucket = %x doesn't exist", []byte(ns))
+		}
+		c := bucket.Cursor()
+		k, v := c.Seek(key)
+		if k == nil || bytes.Compare(k, key) == 1 {
+			k, v = c.Prev()
+			if k == nil || bytes.Compare(k, append(key[:len(key)-8], 0)) <= 0 {
+				// cursor is at the beginning/end of the bucket or smaller than minimum key
+				panic(fmt.Sprintf("BoltDBVersioned.get(), invalid key = %x", key))
+			}
+		}
+		value = make([]byte, len(v))
+		copy(value, v)
+		return nil
+	})
+	if err == nil {
+		return value, nil
+	}
+	if cause := errors.Cause(err); cause == ErrNotExist || cause == ErrBucketNotExist {
+		return nil, err
+	}
+	return nil, errors.Wrap(ErrIO, err.Error())
 }
 
-// Delete deletes a record,if key is nil,this will delete the whole bucket
-func (b *BoltDBVersioned) Delete(ns string, key []byte) error {
+// Delete deletes a record, if key does not exist, it returns nil
+func (b *BoltDBVersioned) Delete(version uint64, ns string, key []byte) error {
 	if !b.db.IsReady() {
 		return ErrDBNotStarted
 	}
-	// TODO: implement Delete
-	return nil
+	// check key's metadata
+	km, err := b.checkNamespaceAndKey(ns, key)
+	if err != nil {
+		if cause := errors.Cause(err); cause != ErrNotExist && cause != ErrInvalid {
+			return err
+		}
+	}
+	if km == nil || version < km.lastVersion || version <= km.deleteVersion {
+		return nil
+	}
+	km.deleteVersion = version
```

Comment:
Only saving the latest delete will cause getting wrong value if multiple deletes. 

for example, there are opt sequnces as follows:
- put k1 v1 at 1
- del k1 at 5
- put k1 v2 at 10
- del k1 at 15

then get k1 at 7 should return NotExist, but actually returns v1